tests/storage/test_state.py

```diff
@@ -483,3 +482,119 @@ def test_get_state_for_event(self):
 
         self.assertEqual(is_all, True)
         self.assertDictEqual({(e5.type, e5.state_key): e5.event_id}, state_dict)
+
+    def test_state_filter_difference(self):
+        def assert_difference(
+            minuend: StateFilter, subtrahend: StateFilter, expected: StateFilter
+        ):
+            self.assertEqual(
+                minuend.approx_difference(subtrahend),
+                expected,
+                f"StateFilter difference not correct:\n\n\t{minuend!r}\nminus\n\t{subtrahend!r}\nwas\n\t{minuend.approx_difference(subtrahend)}\nexpected\n\t{expected}",
+            )
+
+        # it's not possible to subtract individual state keys from
+        # a wildcard
+        assert_difference(
+            StateFilter.all(),
+            StateFilter.from_types(
+                ((EventTypes.Member, "@wombat:hs1"), (EventTypes.Member, "@spqr:hs1"))
+            ),
+            StateFilter.all(),
+        )
+        self.assertEqual(
+            StateFilter.all().approx_difference(
+                StateFilter.from_types(
+                    (
+                        (EventTypes.Member, "@wombat:hs1"),
+                        (EventTypes.Member, "@spqr:hs1"),
+                    )
+                )
+            ),
+            StateFilter.all(),
+        )
+
+        # we can subtract wildcards from wildcards
+        assert_difference(StateFilter.all(), StateFilter.all(), StateFilter.none())
+        assert_difference(
+            StateFilter(
+                types=frozendict(
+                    {
+                        EventTypes.Member: frozenset(),
+                        EventTypes.CanonicalAlias: None,
```

Comment:
Not really relevant to this change, but: it wasn't obvious to me what an empty set and None mean here (and if they ought to have different meanings).

I don't know if there's a nice way. If we had sum types I'd suggest
```rust
enum Fetch {
    All,
    None,
    WithStateKey(set_of_strings),
}
```
But that's slightly awkward because `None` and `WithStateKey(empty_set)` represent the same thing.  /shrug

Reply:
The docstring for StateFilter specifies what they mean, though that's not to say I don't agree with you — and yeah, I think empty set makes sense, but 'None' very much reads as, well, none. It's half tempting to have a constant `WILDCARD` for `None` ;p, but I feel like that might just hide things more.